tonic-web/tonic-web/src/cors.rs

```diff
@@ -0,0 +1,406 @@
+        }
+
+        if !self.is_method_allowed(req_headers.get(REQUEST_METHOD)) {
+            return Err(Error::MethodNotAllowed);
+        }
+
+        let mut headers = self.common_headers(origin.clone());
+        headers.insert(ALLOW_METHODS, self.cache.allow_methods.clone());
+        headers.insert(ALLOW_HEADERS, request_headers_header.clone());
+
+        if let Some(max_age) = self.cache.config.max_age {
+            headers.insert(MAX_AGE, HeaderValue::from(max_age.as_secs()));
+        }
+
+        Ok(headers)
+    }
+
+    pub(crate) fn simple(&self, headers: &HeaderMap) -> Result<HeaderMap, Error> {
+        match headers.get(header::ORIGIN) {
+            Some(origin) if self.is_origin_allowed(origin) => {
+                Ok(self.common_headers(origin.clone()))
+            }
+            Some(_) => Err(Error::OriginNotAllowed),
+            None => Ok(HeaderMap::new()),
+        }
+    }
+
+    fn common_headers(&self, origin: HeaderValue) -> HeaderMap {
+        let mut headers = HeaderMap::new();
+        headers.insert(ALLOW_ORIGIN, origin);
+        headers.insert(EXPOSE_HEADERS, self.cache.expose_headers.clone());
+
+        if self.cache.config.allow_credentials {
+            headers.insert(ALLOW_CREDENTIALS, self.cache.allow_credentials.clone());
+        }
+
+        headers
+    }
+
+    fn is_origin_allowed(&self, origin: &HeaderValue) -> bool {
+        self.cache.config.allowed_origins.is_allowed(origin)
+    }
+
+    #[cfg(test)]
+    pub(crate) fn __check_preflight(&self, headers: &HeaderMap) -> Result<HeaderMap, Error> {
+        self.preflight(
+            headers,
+            headers.get(ORIGIN).unwrap(),
+            headers.get(REQUEST_HEADERS).unwrap(),
+        )
+    }
+}
+
+#[cfg(test)]
+impl Default for Cors {
+    fn default() -> Self {
+        Cors::new(Config::default())
+    }
+}
+
+fn join_header_value<I>(values: I) -> Result<HeaderValue, header::InvalidHeaderValue>
+where
+    I: IntoIterator,
+    I::Item: AsRef<str>,
+{
+    let mut values = values.into_iter();
+    let mut value = Vec::new();
+
+    if let Some(v) = values.next() {
+        value.extend(v.as_ref().as_bytes());
+    }
+    for v in values {
+        value.push(b',');
+        value.extend(v.as_ref().as_bytes());
+    }
+    HeaderValue::from_bytes(&value)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! assert_value_eq {
+        ($header:expr, $expected:expr) => {
+            fn sorted(value: &str) -> Vec<&str> {
+                let mut vec = value.split(",").collect::<Vec<_>>();
+                vec.sort();
+                vec
+            }
+
+            assert_eq!(sorted($header.to_str().unwrap()), sorted($expected))
+        };
+    }
+
+    fn value(s: &str) -> HeaderValue {
+        s.parse().unwrap()
+    }
+
+    impl From<Config> for Cors {
+        fn from(c: Config) -> Self {
+            Cors::new(c)
+        }
+    }
+
+    #[test]
+    #[should_panic]
+    #[ignore]
```

Comment:
why is this ignored?

Reply:
Because `allow_origins` is supposed to check that the values are valid values but I have not implemented that check yet. Right now it only checks that are valid `HeaderValue`s but not valid origins.